rust/user-info-fetcher/src/backend/active_directory.rs

```diff
@@ -0,0 +1,394 @@
+
+    #[snafu(display("failed to connect to LDAP"))]
+    ConnectLdap { source: LdapError },
+
+    #[snafu(display("failed to send LDAP request"))]
+    RequestLdap { source: LdapError },
+
+    #[snafu(display("failed to bind LDAP credentials"))]
+    BindLdap { source: LdapError },
+
+    #[snafu(display("failed to search LDAP for users"))]
+    FindUserLdap { source: LdapError },
+
+    #[snafu(display("failed to search LDAP for groups of user"))]
+    FindUserGroupsLdap { source: LdapError },
+
+    #[snafu(display("invalid user ID sent by client"))]
+    ParseIdByClient { source: uuid::Error },
+
+    #[snafu(display("invalid user ID sent by LDAP"))]
+    ParseIdByLdap { source: uuid::Error },
+
+    #[snafu(display("unable to find user {request}"))]
+    UserNotFound { request: ErrorRenderUserInfoRequest },
+
+    #[snafu(display("unable to parse user {user_dn:?}'s primary group's RID"))]
+    InvalidPrimaryGroupRelativeId {
+        source: ParseIntError,
+        user_dn: String,
+    },
+
+    #[snafu(display("user {user_dn:?}'s SID has no subauthorities"))]
+    UserSidHasNoSubauthorities { user_dn: String },
+
+    #[snafu(display("failed to parse user {user_dn:?}'s SID"))]
+    ParseUserSid {
+        source: ParseSecurityIdError,
+        user_dn: String,
+    },
+}
+
+impl http_error::Error for Error {
+    fn status_code(&self) -> StatusCode {
+        match *self {
+            Error::ConfigureTls { .. } => StatusCode::SERVICE_UNAVAILABLE,
+            Error::ConnectLdap { .. } => StatusCode::SERVICE_UNAVAILABLE,
+            Error::RequestLdap { .. } => StatusCode::SERVICE_UNAVAILABLE,
+            Error::BindLdap { .. } => StatusCode::SERVICE_UNAVAILABLE,
+            Error::FindUserLdap { .. } => StatusCode::SERVICE_UNAVAILABLE,
+            Error::FindUserGroupsLdap { .. } => StatusCode::SERVICE_UNAVAILABLE,
+            Error::ParseIdByClient { .. } => StatusCode::BAD_REQUEST,
+            Error::ParseIdByLdap { .. } => StatusCode::INTERNAL_SERVER_ERROR,
+            Error::UserNotFound { .. } => StatusCode::NOT_FOUND,
+            Error::InvalidPrimaryGroupRelativeId { .. } => StatusCode::INTERNAL_SERVER_ERROR,
+            Error::UserSidHasNoSubauthorities { .. } => StatusCode::INTERNAL_SERVER_ERROR,
+            Error::ParseUserSid { .. } => StatusCode::INTERNAL_SERVER_ERROR,
+        }
+    }
+}
+
+// Matching rules defined at https://learn.microsoft.com/en-us/windows/win32/adsi/search-filter-syntax#operators
+/// Makes DN filters apply recursively to group membership
+const LDAP_MATCHING_RULE_IN_CHAIN: &str = ":1.2.840.113556.1.4.1941:";
+
+const LDAP_FIELD_OBJECT_ID: &str = "objectGUID";
+const LDAP_FIELD_OBJECT_SECURITY_ID: &str = "objectSid";
+const LDAP_FIELD_OBJECT_DISTINGUISHED_NAME: &str = "dn";
+const LDAP_FIELD_USER_NAME: &str = "userPrincipalName";
+const LDAP_FIELD_USER_PRIMARY_GROUP_RID: &str = "primaryGroupID";
+const LDAP_FIELD_GROUP_MEMBER: &str = "member";
+
+pub(crate) async fn get_user_info(
```

Comment:
Is this not instrumented for tracing on purpose?

Reply:
No good reason, that makes sense to add.

Reply:
3fc772a478228e120a65ef5058beda8385151eaa